File: stacks-and-queues/3.2.py
class StackWithMin:
    def __init__(self):
        self.data = []
        self.mins = []

    def push(self, val):
        if len(self.mins) == 0 or val <= self.mins(-1):
            self.mins.append(val)
        self.data.append(val)

    def pop(self):
        if self.data[-1] == self.mins[-1]:
            self.mins.pop()
        return self.data.pop()

    def min(self):
        return self.mins[-1]

    def peek(self):
        return self.data[-1]

File: stacks-and-queues/3.2.rb
class StackWithMin
  def initialize
    @data = []
    @mins = []
  end

  def push(val)
    if @mins.size == 0 || val <= @mins.last
      @mins.push val
    end

    @data.push val
  end

  def pop
    @mins.pop if @mins.last == @data.last
    @data.pop
  end

  def min
    @mins.last
  end

  def peek
    @data.last
  end
end
